File: CsharpProjects/myLeetCode/Dynamic_Programming/Basic/746. Min Cost Climbing Stairs.cs
// LeetCode 746. Min Cost Climbing Stairs
// https://leetcode.com/problems/min-cost-climbing-stairs/

#include <iostream>
#include <vector>

int minCostClimbingStairs(std::vector<int>& cost) {
    int n = cost.size();
    std::vector<int> dp(n+1, 0);
    dp[0] = 0;
    dp[1] = 0;

    for (int i = 2;i < n+1;++i) {
        dp[i] = std::min(dp[i-1]+cost[i-1],dp[i-2]+cost[i-2]);
    }

    return dp[n];
}

int main() {
    std::vector<std::vector<int>> testCases = {{10, 15, 20}, {1, 100, 1, 1, 1, 100, 1, 1, 100, 1}};
    for (auto& testCase: testCases) {
        std::cout << "The minimum cost to reach the top is: " << minCostClimbingStairs(testCase) << std::endl;
    }
    
    return 0;
}

File: C++/代码随想录/Dynamic_Programming/Basic/746. Min Cost Climbing Stairs.py
def minCostClimbingStairs(cost):
    n = len(cost)
    if n == 0 or n == 1:
        return 0

    min_cost = [0] * (n + 1)
    min_cost[1] = cost[0]

    for i in range(2, n + 1):
        min_cost[i] = min(min_cost[i - 1], min_cost[i - 2]) + (cost[i - 1] if i < n else 0)

    return min_cost[n]

# 测试用例
print(minCostClimbingStairs([10, 15, 20]))  # 输出应该是 15
print(minCostClimbingStairs([1, 100, 1, 1, 1, 100, 1, 1, 100, 1]))  # 输出应该是 6
